qiskit_experiments/data_processing/data_processor.py

```diff
@@ -0,0 +1,118 @@
+
+from qiskit_experiments.data_processing.data_action import DataAction
+from qiskit_experiments.data_processing.exceptions import DataProcessorError
+
+
+class DataProcessor:
+    """
+    A DataProcessor defines a sequence of operations to perform on experimental data.
+    Calling an instance of DataProcessor applies this sequence on the input argument.
+    A DataProcessor is created with a list of DataAction instances. Each DataAction
+    applies its _process method on the data and returns the processed data. The nodes
+    in the DataProcessor may also perform data validation and some minor formatting.
+    The output of one data action serves as input for the next data action.
+    DataProcessor.__call__(datum) usually takes in an entry from the data property of
+    an ExperimentData object (i.e. a dict containing metadata and memory keys and
+    possibly counts, like the Result.data property) and produces the formatted data.
+    DataProcessor.__call__(datum) extracts the data from the given datum under
+    DataProcessor._input_key (which is specified at initialization) of the given datum.
+    """
+
+    def __init__(self, input_key: str, data_actions: List[DataAction] = None):
+        """Create a chain of data processing actions.
+
+        Args:
+            input_key: The initial key in the datum Dict[str, Any] under which the data processor
+                will find the data to process.
+            data_actions: A list of data processing actions to construct this data processor with.
+                If None is given an empty DataProcessor will be created.
+        """
+        self._input_key = input_key
+        self._nodes = []
+
+        if data_actions:
+            for node in data_actions:
+                self.append(node)
+
+    def append(self, node: DataAction):
+        """
+        Append new data action node to this data processor.
+
+        Args:
+            node: A DataAction that will process the data.
+        """
+        self._nodes.append(node)
+
+    def __call__(self, datum: Dict[str, Any]) -> Tuple[Any, Any]:
+        """
+        Call self on the given datum. This method sequentially calls the stored data actions
+        on the datum.
+
+        Args:
+            datum: A single item of data, typically from an ExperimentData instance, that needs
+                to be processed. This dict also contains the metadata of each experiment.
+
+        Returns:
+            processed data: The data processed by the data processor.
+
+        Raises:
+            DataProcessorError: if no nodes are present.
+        """
+        if len(self._nodes) == 0:
+            raise DataProcessorError("Cannot call an empty data processor.")
+
+        datum_ = datum[self._input_key]
+
+        for node in self._nodes:
+            datum_ = node(datum_)
+
+        return datum_
+
+    def call_with_history(
+        self, datum: Dict[str, Any], history_nodes: Set = None
```

Comment:
After much discussion, we decided to separate to two functions, with and without history. Still they could be implemented such that they both call one internal function with a flag.

Reply:
This can change the signature of function depends on flag, which makes hard to understand the behavior of call without reading the code.

Reply:
This can change the signature of function depends on flag, which makes hard to understand the behavior of call without reading the code.

Reply:
I still think the duplication here should be resolved. For example, if one day you want to change something, you'll have to remember to make the change in both functions.

Reply:
Do you have any suggestions as to how one would do that while still being efficient?

Reply:
See here: https://github.com/Qiskit/qiskit-experiments/pull/22/files#r616574994

Reply:
Thank you. See here: bc00e26